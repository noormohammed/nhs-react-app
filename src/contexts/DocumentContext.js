import React, {
  createContext,
  useContext,
  useState,
  useEffect,
  useRef,
} from "react";

/**
 * DocumentContext for Document details
 */

const DocumentContext = createContext({});

/**
 * Document Provider using the DocumentContext
 * @param {*} children any props/children components passed to the DocumentContextProvider
 * @returns memoized object of document info/details state, set state function
 */
const DocumentContextProvider = ({ children }) => {
  // Once the document info is memoized, we store the info in the localStorage
  // for ease of access and its also economical (can avoid too many backend calls).
  const [documentInfo, setDocumentInfo] = useState(
    window.localStorage.getItem("document_info")
      ? JSON.parse(window.localStorage.getItem("document_info"))
      : null
  );

  const isMountedRef = useRef(null);

  useEffect(() => {
    isMountedRef.current = true;
    const storeDocumentInfo = async () => {
      if (isMountedRef.current) {
        window.localStorage.setItem("document_info", JSON.stringify(documentInfo));
      }
    };
    storeDocumentInfo();

    return () => {
      isMountedRef.current = false;
    };
  }, [documentInfo]);

  return (
    <DocumentContext.Provider value={[documentInfo, setDocumentInfo]}>
      {children}
    </DocumentContext.Provider>
  );
};

const useDocumentContext = () => useContext(DocumentContext);

export { useDocumentContext, DocumentContextProvider };
