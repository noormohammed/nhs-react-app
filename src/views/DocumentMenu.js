import React, { useEffect } from "react";
import { useHistory } from "react-router-dom";

import { useDocumentContext } from "contexts/DocumentContext";
import Page from "components/Page";

const DocumentMenu = () => {
  const history = useHistory();
  const [documentInfo, setDocumentInfo] = useDocumentContext();

  useEffect(() => {
    setDocumentInfo({
      ...documentInfo,
      mode: "menu",
    });
    return () => {};
  }, []);

  return (
    <Page title="Document Menu">
      <div className="document">
        <div className="heading">
          Below are the options available for this document.
        </div>

        <div className="table">
          <div className="table-body">
            <div className="row">
              <div className="table-header label column column-heading">
                MENU OPTIONS
              </div>
              <div className="column">
                Edit
                <button
                  className="pull-right"
                  onClick={() => {
                    history.push("/document-edit");
                  }}
                >
                  <span className="caret right"></span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </Page>
  );
};

export default DocumentMenu;
