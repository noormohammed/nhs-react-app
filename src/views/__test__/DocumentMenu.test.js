import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createBrowserHistory } from "history";
import { Router } from "react-router-dom";

import { DocumentContextProvider } from "contexts/DocumentContext";
import DocumentMenu from "views/DocumentMenu";

const browserHistory = createBrowserHistory();

const documentInfo = null;
const setDocumentInfo = jest.fn();

describe("<DocumentMenu /> Tests", () => {
  let container;
  beforeEach(() => {
    ({ container } = render(
      <DocumentContextProvider value={[documentInfo, setDocumentInfo]}>
        <Router history={browserHistory}>
          <DocumentMenu />
        </Router>
      </DocumentContextProvider>
    ));
  });

  afterEach(cleanup);

  test("renders DocumentMenu correctly", () => {
    expect(container).toMatchSnapshot();
  });

  test("renders correct title", () => {
    expect(screen.getByText(/Document Menu/i)).toBeTruthy();
  });

  test("Menu Options has an option called 'Edit'", () => {
    expect(screen.getByText(/Edit/i)).toBeInTheDocument();
  });

  test("Menu Options has an Edit action button", () => {
    const editBtn = container.getElementsByClassName("menu-action-button");
    expect(editBtn.length).toEqual(1);
  });

  test("trigger Edit action button click to push Edit Document view", () => {
    userEvent.click(screen.getByTestId("edit-action-button"));
    expect(browserHistory.location.pathname).toBe("/document-edit");
  });
});
