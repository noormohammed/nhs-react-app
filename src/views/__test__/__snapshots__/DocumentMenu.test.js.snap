// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DocumentMenu /> Tests renders DocumentMenu correctly 1`] = `
<div>
  <div
    class="main"
  >
    <div
      style="position: fixed; top: 0px; display: flex; flex-direction: row; justify-content: center; align-items: center; width: 100%; height: 50px; background-color: rgb(63, 80, 181); border-bottom: 1px solid #d8d8d8; font-weight: bold; padding: 0px 20px 0px 0px; box-sizing: border-box; color: rgb(255, 255, 255); font-size: 18px;"
    >
      <div
        style="width: 50px;"
      >
        <button>
          ⬅️
        </button>
      </div>
      <div
        style="margin-left: auto; margin-right: auto;"
      >
        Document Menu
      </div>
      <div
        style="width: 50px;"
      />
    </div>
    <div
      class="content"
    >
      <div
        class="document"
      >
        <div
          class="heading"
        >
          Below are the options available for this document.
        </div>
        <div
          class="table"
        >
          <div
            class="table-body"
          >
            <div
              class="row"
            >
              <div
                class="table-header label column column-heading"
              >
                MENU OPTIONS
              </div>
              <div
                class="column"
              >
                <button
                  class="menu-action-button"
                  data-testid="edit-action-button"
                >
                  Edit
                  <span
                    class="caret right"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="display: flex; align-items: center; justify-content: center; width: 100%; height: 40px; background-color: rgb(51, 51, 51); color: rgb(255, 255, 255); position: fixed; bottom: 0px;"
    >
      Footer Menu
    </div>
  </div>
</div>
`;
